libs/vkd3d/memory.c

```diff
@@ -0,0 +1,482 @@
+{
+    if (device->device_info.buffer_device_address_features.bufferDeviceAddress)
+        allocation->resource.va = vkd3d_get_buffer_device_address(device, allocation->resource.vk_buffer);
+    else
+    {
+        /* TODO implement */
+    }
+
+    /* TODO register GPU address */
+
+    return S_OK;
+}
+
+static void vkd3d_memory_allocation_free(const struct vkd3d_memory_allocation *allocation, struct d3d12_device *device, struct vkd3d_memory_allocator *allocator)
+{
+    const struct vkd3d_vk_device_procs *vk_procs = &device->vk_procs;
+
+    /* TODO unregister GPU address */
+
+    if (allocation->flags & VKD3D_ALLOCATION_FLAG_GLOBAL_BUFFER)
+        VK_CALL(vkDestroyBuffer(device->vk_device, allocation->resource.vk_buffer, NULL));
+
+    VK_CALL(vkFreeMemory(device->vk_device, allocation->vk_memory, NULL));
+}
+
+static HRESULT vkd3d_memory_allocation_init(struct vkd3d_memory_allocation *allocation, struct d3d12_device *device,
+        struct vkd3d_memory_allocator *allocator, const struct vkd3d_allocate_memory_info *info)
+{
+    const struct vkd3d_vk_device_procs *vk_procs = &device->vk_procs;
+    VkMemoryRequirements memory_requirements;
+    VkMemoryAllocateFlagsInfo flags_info;
+    VkMemoryPropertyFlags type_flags;
+    uint32_t type_mask;
+    VkResult vr;
+    HRESULT hr;
+
+    memset(allocation, 0, sizeof(*allocation));
+    allocation->heap_type = info->heap_properties.Type;
+    allocation->heap_flags = info->heap_flags;
+    allocation->flags = info->flags;
+
+    /* This also sort of validates the heap description,
+     * so we want to do this before creating any objects */
+    if (FAILED(hr = vkd3d_select_memory_flags(device, &info->heap_properties, &type_flags)))
+        return hr;
+
+    if (allocation->flags & VKD3D_ALLOCATION_FLAG_GLOBAL_BUFFER)
+    {
+        /* If requested, create a buffer covering the entire allocation
+         * and derive the exact memory requirements from that. Any buffer
+         * resources are just going to use this buffer with an offset. */
+        if (FAILED(hr = vkd3d_create_global_buffer(device, info->memory_requirements.size,
+                &info->heap_properties, info->heap_flags, &allocation->resource.vk_buffer)))
+            return hr;
+
+        VK_CALL(vkGetBufferMemoryRequirements(device->vk_device,
+                allocation->resource.vk_buffer, &memory_requirements));
+
+        memory_requirements.memoryTypeBits &= info->memory_requirements.memoryTypeBits;
+    }
+    else
+    {
+        /* Respect existing memory requirements since there may not
+         * be any buffer resource to get memory requirements from. */
+        memory_requirements = info->memory_requirements;
+    }
+
+    /* For dedicated buffer allocations we should assign the existing
+     * buffer for address lookup purposes, but take care not to destroy
+     * it when freeing the allocation. */
+    if (allocation->flags & VKD3D_ALLOCATION_FLAG_DEDICATED_BUFFER)
```

Comment:
I guess DEDICATED and GLOBAL cannot be used together here, or we leak allocation->resource.vk_buffer?

Reply:
Yes, they are mutually exclusive, see above.